src/Markup/Avalonia.Markup/Data/ExpressionNode.cs

```diff
@@ -2,124 +2,188 @@
+                    _valueSubscription = null;
+                    _valuePluginSubscription?.Dispose();
                     _target = value;
 
-                    if (newInstance != null)
+                    if (running)
                     {
-                        SubscribeAndUpdate(_target);
-                    }
-                    else
-                    {
-                        CurrentValue = UnsetReference;
-                    }
-
-                    if (Next != null)
-                    {
-                        Next.Target = _value;
+                        _valueSubscription = StartListeningCore();
                     }
                 }
             }
         }
 
-        public WeakReference CurrentValue
+        public IDisposable Subscribe(IObserver<object> observer)
         {
-            get
+            if (_observer != null)
             {
-                return _value;
+                throw new AvaloniaInternalException("ExpressionNode can only be subscribed once.");
             }
 
-            set
+            _observer = observer;
+            var nextSubscription = Next?.Subscribe(this);
+            _valueSubscription = StartListeningCore();
+
+            return Disposable.Create(() =>
             {
-                _value = value;
+                _valueSubscription?.Dispose();
+                _valueSubscription = null;
+                _valuePluginSubscription?.Dispose();
+                _valuePluginSubscription = null;
+                nextSubscription?.Dispose();
+                _observer = null;
+            });
+        }
 
-                if (Next != null)
-                {
-                    Next.Target = value;
-                }
+        void IObserver<object>.OnCompleted()
+        {
+            throw new AvaloniaInternalException("ExpressionNode.OnCompleted should not be called.");
+        }
 
-                _subject?.OnNext(value.Target);
-            }
+        void IObserver<object>.OnError(Exception error)
+        {
+            throw new AvaloniaInternalException("ExpressionNode.OnError should not be called.");
         }
 
-        public virtual bool SetValue(object value, BindingPriority priority)
+        void IObserver<object>.OnNext(object value)
         {
-            return Next?.SetValue(value, priority) ?? false;
+            NextValueChanged(value);
         }
 
-        public virtual IDisposable Subscribe(IObserver<object> observer)
+        protected virtual IObservable<object> StartListening(WeakReference reference)
         {
-            if (Next != null)
+            return Observable.Return(reference.Target);
+        }
+
+        protected virtual void NextValueChanged(object value)
+        {
+            var bindingBroken = BindingNotification.ExtractError(value) as MarkupBindingChainNullException;
+            bindingBroken?.AddNode(Description);
+            _observer.OnNext(value);
+        }
+
+        private IDisposable StartListeningCore()
```

Comment:
I think that the names of `StartListeningCore` and `StartListening` should be swapped. Method names ending with `Core` should be the subclass implementation.